include/irr/asset/IDescriptorSet.h

```diff
@@ -60,10 +61,26 @@ class IDescriptorSet
     };
 
 protected:
-    IDescriptorSet(core::smart_refctd_dynamic_array<SWriteDescriptorSet>&& _descriptors) : m_descriptors(std::move(_descriptors)) {}
+    /**
+    @param _descriptors Entries must be sorted by binding number
+    */
+    IDescriptorSet(core::smart_refctd_dynamic_array<SWriteDescriptorSet>&& _descriptors, core::smart_refctd_dynamic_array<uint32_t>&& _dynOffsets) : 
+        m_descriptors(std::move(_descriptors)), m_dynamicOffsets(std::move(_dynOffsets)) 
+    {
+        auto is_not_sorted = [this] {
+            for (auto it = m_descriptors->cbegin()+1; it != m_descriptors->cend(); ++it)
+                if (it->binding <= (it-1)->binding)
+                    return false;
+            return true;
+        };
+        auto is_dynamic = [] (const SWriteDescriptorSet& _entry) { return _entry.descriptorType==EDT_UNIFORM_BUFFER_DYNAMIC || _entry.descriptorType==EDT_STORAGE_BUFFER_DYNAMIC; };
+        assert(std::count_if(m_descriptors->cbegin(), m_descriptors->end(), is_dynamic) == m_dynamicOffsets->size());
+        assert(!is_not_sorted);
+    }
     virtual ~IDescriptorSet() = default;
 
     core::smart_refctd_dynamic_array<SWriteDescriptorSet> m_descriptors;
+    core::smart_refctd_dynamic_array<uint32_t> m_dynamicOffsets;
```

Comment:
no, no, no, no, no

dynamic offset is NOT part of the descriptor set state

you pass dynamic offsets in when you bind descriptor set in a command buffer (IVideoDriver for now)

Reply:
@Crisspl !?

Reply:
yea, but user doesn't bind descriptor sets. They're taken from 4 places in the engine (global,camera,scenenode,meshbuffer) and bound automatically. That's why i came up with an idea to have dynamic offsets in IDescriptorSet

Reply:
ofc user can bind descirptor set
`IVideoDriver::bindDescriptorSets()`
would have to be indentical to command buffer op in vulkan
https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkCmdBindDescriptorSets.html

Reply:
ok now im confused about the design

Reply:
`IVideoDriver::setDescriptorSets` needs to take a(n):
- enum graphics vs compute
- pipeline layout
- first desc set ix
- desc set count
- up to 4 desc sets
- dynamic offset count
- dynamic offset array pointer

Excerpt from spec:

>If any of the sets being bound include dynamic uniform or storage buffers, then pDynamicOffsets includes one element for each array element in each dynamic descriptor type binding in each set. Values are taken from pDynamicOffsets in an order such that all entries for set N come before set N+1; within a set, entries are ordered by the binding numbers in the descriptor set layouts; and within a binding array, elements are in order. dynamicOffsetCount must equal the total number of dynamic descriptors in the sets being bound.

> The effective offset used for dynamic uniform and storage buffer bindings is the sum of the relative offset taken from pDynamicOffsets, and the base address of the buffer plus base offset in the descriptor set. The range of the dynamic uniform and storage buffer bindings is the buffer range as specified in the descriptor set.

> Each of the pDescriptorSets must be compatible with the pipeline layout specified by layout. The layout used to program the bindings must also be compatible with the pipeline used in subsequent graphics or compute commands, as defined in the Pipeline Layout Compatibility section.

Reply:
yea, but now i dont see point of those descriptor sets in [global],camera,scenenode,meshbuf if user has API to bind them directly

Reply:
the point of 
> descriptor sets in [global],camera,scenenode,meshbuf
is to make the `scene` namespace usable

`video` namespace must be independent of `scene`

this  is so we can have things like instanved scene nodes while having total access to the OpenGL 4.3-4.6 and Vulkan API ( have our cake and eat it)

Reply:
scene namespace requires video namespace, so it is usable without "[global],camera,scenenode,meshbuf" if there is API to bind descriptor sets.
Also: how engine decides which descriptor sets to bind? those from scenenode, meshbuffer, etc.. or the ones explicitly bound by user with API?

Reply:
just like `setMaterial`
SceneNode::render() sets the `setMaterial` by default, if user comes in and does their own `setMaterial` then we use that.

Reply:
"[global],camera,scenenode,meshbuf" this stuff I only want to be considered at the `scene` namespace level.

Remember `SMaterial` in meshbuffers? nothing in `video` set the material from meshbuffer for you!

`setMaterial` had to be called before `draw`, `draw` did not automagically set the material for you from the meshbuffer, even though the meshbuffer had it!

Reply:
@Crisspl can we implement 
https://github.com/buildaworldnet/IrrlichtBAW/pull/358#discussion_r328371583
properly for the pipelines?

Reply:
@Crisspl this is resolved, but useful to review for the two biggest unresolves in the whole PR.

Moving conversation to #376